### ISSUE TITLE
OreGen / WorldGen

### ISSUE THREAD

**Issue by magik6k** [NOVAAPI/NovaSuggestions#15]
Any ideas on portable design of that?

**Comment by stanhebben** [comment]
A WorldGenerator class or interface could be implemented and registered by any mod which wishes to do worldgen. This generator could receive a BlockAccess to work on, as well as a region (x/y/z coordinate range) that needs to be populated.

For Minecraft, this would like be called one chunk at a time, but other games could call it with different regions.

**Comment by stanhebben** [comment]
Note from Cervator: if you're curious to investigate an alternative approach you can check out https://github.com/Terasology/TutorialWorldGeneration/wiki for an intro to our world gen, which i suspect differs a bit from standard MC. Also helps highlight the challenge of writing a wrapper supporting multiple approaches :-)

**Comment by NotGyro** [comment]
A couple of thoughts - 
1. World generation as an optional API? Some voxel games may have pre-made levels.
2. Ores in stones - perhaps an OreAPI designed for the specific purpose of transposing modded ore textures with transparency onto the base game (or another mod)'s stones? Nobody wants to see an out-of-place Minecraft smooth stone texture in the middle of rocks in Terasology, for example.

**Comment by keybounce** [comment]
Something that would be nice, but probably hard: letting world generators be called by other world generators. In other words, isolating and separating "Here is a chunk that was asked for" from "This chunk is now added to the world".

The intended use is to permit a world generator to work by combining operations from multiple world generators -- so for example, asking the nether generator for a chunk, stripping everything above Y=120, asking the overworld generator for a chunk, stripping everything below Y=35, and then merging the two (letting the air between overworld Y=35 and Y=45 carve holes in the nether Y=110 to Y=120), and stacking them on top of each other. The result is an "overworld-style" top, and a "nether-style" bottom -- a "Lost worlds" type of thing, with intersections in the deep oceans.

For vanilla minecraft, this is difficult, at best. The only way I can see to do this is to have the various generators called for temporary dimensions, such as -129 and -130 (numbers specifically chosen to be outside of the map/login dimension zone), then combine that into a final chunk that is added to the target world. The issue is that chunk decorations (at least) can trigger generation of other chunks.

Note that getting raw chunks, and getting decorated chunks, are two different issues. Being able to get both chunks decorated -- so that mineshafts are in the proper place in the upper world, caves run throughout, etc -- is most definitely non-trivial, as any chunk can be decorated by a later chunk (mineshafts, giant trees), and snow has to be placed down after nearby chunks have decorated this chunk. 

(Additionally, both COG and Mystcraft have had to determine how to tell when nearby chunks have finished decorating so that they can do things -- so this is becoming a more general issue, and ignored by Forge).

**Comment by NotGyro** [comment]
...Actually, there should be a more generalized OverlayAPI for placing modded overlays on the base game's textures. Not sure about the details, though.

**Comment by tattyseal** [comment]
Do you think there should be an OreGenerator that is basic and has a registry which takes rarity, cluster size etc. and a separate AdvancedWorldGenerator that can generate ores as well as structures in 1?

**Comment by stanhebben** [comment]
I think an Ore generation API would be nice to have, provides users with control over ore generation (where possible). For games with static worlds, world generators would simply never be called - technically the world is already generated.

I think that ore generation and structure generation are two specializations of the same concept - worldgen - and that they could both be implemented with the same system.

**Comment by Kubuxu** [comment]
Idea for ore generation is to work on relative (game independent) values. Depth of 20 means nothing if world spans from -2^31 to 2^31 and as we plan making this as universal as possible it is important to take it into account.

**Comment by tattyseal** [comment]
@Kubuxu Currently it will work as 0 is ground level and then -1f will be relative to that games depth, in minecrafts case 32.

**Comment by keybounce** [comment]
But how do you handle worlds that are taller?

For example: Minecraft 0=sea level (y=62), then -1F (bedrock) would only be around the tops of extreme hills, and less than savannah M (or whichever was that crazy cloud-level height).

And that's before an amplified world makes Y=200 all over the place.

**Comment by NotGyro** [comment]
A generalized oregen module would help avoid a bunch of code duplication so yeah, good idea. Especially if different wrappers can handle it differently.

One way to do it for any world height would be to allow the developer to specify things like "from minimum possible height to (sea level + X)". That would be pretty neat. Alternatively, you could provide the mod developers with a way to query the height range of the world, and then they can choose heights to generate things based upon that or based upon whatever whim they have at the moment.

**Comment by calclavia** [comment]
Yes. World.getHeight()

**Comment by LudziE12** [comment]
Yes, there should be support for worlds with 3d chunks, i mean y = -100 and deeper.

**Comment by keybounce** [comment]
Alright, how would you handle arbitrary depths?

Lets say you have a cubic chunks game, and a world generator that will usually make heights in the sea level+25 range (say, minecraft hills), occasionally in the sea level+100 range (extreme hills, savannah M), but has no inherent upper limit -- perhaps it is similar to the cubic chunk proof of concept video where terrain was going into the Y=1,000 range, perhaps occasionally you find Himalaya-style mountains.

Or, a world that has "all the ground you want to dig in".

What if you had a mod like Custom Ore Gen, that did something like "this ore is found just below the surface, this ore is found half-way to bedrock, this ore is found down at the lava lake level, this ore is found down at the bedrock level, this ore is found up in the surface hills".

**Comment by calclavia** [comment]
Use percentage values. -1 is the maximum possible depth. 0 is sea level. 1 is maximum height.

**Comment by Kubuxu** [comment]
@calclavia I don't think that using -1 for maxmum possible depth is good idea. I would rather say -1 as a average depth of universal deposits. Like Iron. It would make support for games with large heights and depth much more easier.

**Comment by NotGyro** [comment]
Maybe the developer could input how many standard deviations away from the mean this ore generates, out of all ores generating in the world? Or would that be too complicated / fucky?

**Comment by Kubuxu** [comment]
I would say. Maybe we could standardize depths for minerals (-1 for iron like, -2 for gold like and -3 for diamond like) and allow wrapper provider to transform them into game's depths with usage for functional interface.

**Comment by immibis** [comment]
What would games without ores do?

**Comment by Kubuxu** [comment]
We could make this module optional as we want the game to generate ores in game specific way.

**Comment by keybounce** [comment]
On 2015-02-01, at 9:18 PM, Jakub Sztandera <notifications@github.com> wrote:

> I would say. Maybe we could standardize depths for minerals (-1 for iron like, -2 for gold like and -3 for diamond like) and allow wrapper provider to transform them into game's depths with usage for functional interface.
> 
Bad idea. And, the core of a good idea.

Lets say you have a system of veins. Lets say your redstone veins run vertically from slightly higher than sea level down to about half the typical ground depth, and your lapis veins run vertically from about the ocean bottom down to the bedrock area.

Or, lets say you have something that runs in three diagonally sloping down lines from a common point.
Or, something that runs in veins high up in the mountains.

COG is probably the best example of "This breaks my assumptions" because it doesn't have ores everwhere. The mod support for different mod ores uses some different distributions for each, based on a normal vanilla world of "About 60 blocks under the sea level, about 5 blocks above sea level in most areas, and different things in mountains".

It's pretty easy to fix the two oddballs in the standard distribution (redstone and lapis) to work properly in twilight forest (half the ground); the rest work just fine, just squished closer to the surface. But looking at really deep underworlds makes the whole thing change.

And, Forestry, probably others, do change how their ores go -- sometimes only above a certain height (instead of below), sometimes around the stone/dirt boundary, etc.

Still, there might be something to this. For vanilla, for example, there is a clear set of "lines in the ground", at (these are from memory) 59 (where iron generates below), 30 (where the rich stuff generates below), and ... 12? (where diamond/redstone generate below). (They all get moved a bit at random, so they can appear above those numbers occasionally). So if you had a good indication of "ground level", "sea level", "basic minerals", "Advanced minerals", and "end-game minerals", you might be able to do something from that.

"Ground level" is a concept that does not exist in vanilla. ATG is the only worldgen I know of that provides that information to mods. COG does try to guess at it, but gets it wrong. And even then, trying to figure out what the ground level is like in general is painful  (I'm thinking of Vanilla's stone noise map, where the height variation does all sorts of contortions -- and that's before something like Mesa Bryce adds in additional ground height during the decoration phase that isn't even tracked by the biome height data).

---

And if you really want a challenge, look at the Geologica config for placement of things. It's an attempt by someone that knows geology to actually try to place ground rocks and ores in a realistic manner, with different rules for strata layers and rock formations in ocean biomes, volcanic biomes, and normal biomes.
Figure out how to map that to a set of numbers, along with some "basic typical ores", and you might have something.

But again: trying to make this general will need some sense of "we know the typical ground level is X above the sea level; we know the typical stone starts at Y above the sea level". Vanilla has some good values for most of the biomes because most of them are typically in the 64-70 range. Once you get into general voxels, or strange worldgen (ATG, RWG, amplified, etc.), things change.=

**Comment by Kubuxu** [comment]
You must remember we are not working only for minecraft here. Imagine game which deposits are on depth of 200m-3km. We need general solution. And I think we should first support game specific type of ore generation. One game might generate ore in groups of 5 blocks, like MC, but other might make enormous veins in depths of world.

**Comment by asiekierka** [comment]
You must also remember that abstracting things too far and leaving too much to the wrapper kills creativity.

**Comment by Kubuxu** [comment]
Creativity is fine but when you want to generate ore in weird form, you are not seeking for ore generation but for structure generation. I would much prefer ores in games with TerraFirma like generation to not to generate like in minecraft.

**Comment by keybounce** [comment]
What was TFC ore generation like then? (I never played with it)

**Comment by magik6k** [comment]
It had hudge, rare ores, deeply under ground. Note games that want to generate ores like coal at levels like -500 blocks

**Comment by keybounce** [comment]
Huge, rare ores? That's kinda like the COG system. But even then, COG veins put iron at around 40-60 (so about the 25 blocks below sea level), or gold around 25-45 (about 20-40 blocks below sea level). (Yes, there's quite a bit of variation). Not 500 blocks down.

(All those heights are scaled by the world's sea level, so they would be closer to the surface in twilight forest).

**Comment by magik6k** [comment]
I'd propose some kind of WorldProperties class that'd provide variables such as oreScale, oreFrequency, seaLevel, highOreLevel, deepOreLevel so generator knows how to distribute ores

**Comment by Kubuxu** [comment]
@keybounce imagine game where you don't have arbitrary 256 height limit. What would you do with all this space? I would fill it with enormous mountains and extremely deep caves. I would give a prize for finding deposit of ore.

**Comment by keybounce** [comment]
Almost forgot -- COG also has "cloud" generation. Think large, but sparse (10-30% coverage) ellipsoid, stretched over several chunks, almost always in the middle of the ground (say, centered around 20-45). Different ores have different frequency, biomes, and density, more than they have different sizes (but notably, emeralds are thin, almost flat, and significantly less wide; diamonds are smaller, deeper, and rarer; etc.).

---

Treating these types of ores as structure gen? ... Technically, they are, but they are not the same as "wizard towers", or "farm building". They would be used in worlds that have ores, even if those worlds don't have other structures.

**Comment by magik6k** [comment]
TerraFirmaCraft oregen for reference: http://wiki.terrafirmacraft.com/Ores_&_Minerals

**Comment by keybounce** [comment]
> @keybounce imagine game where you don't have arbitrary 256 height limit. What would you do with all this space? I would fill it with enormous mountains and extremely deep caves. I would give a prize for finding deposit of ore.

That's a really good question. I have no idea how to adapt COG to something like CubicChunks where the ground goes deep underground. Heck, I am still trying to figure out how to deal with COG and that "lost world" type of worldgen that I mentioned before, let alone with a normal "nether" style world.

If the world could go down 1000 blocks, instead of only down 60-ish blocks, would you want to have to go 900 down before you had a chance to find diamonds?

**Comment by magik6k** [comment]
@keybounce to be honest, totally yes

**Comment by keybounce** [comment]
Then, you'd love the standard COG distributions. You'd probably get hundreds of blocks of iron veins, just rare (they are not more frequent, but there's more ground for them to hide in), then hundreds of blocks of gold (ditto), and finally, near the bottom, kimberlite lava pipes with diamonds.

**Comment by tattyseal** [comment]
@calclavia Floating Point values is what I was going for. It would give the mod developer much more control like you said, 1 is max height, -1 min height and 0 ground level. But ofcourse try and do it per chunk so 0 is always the ground level of that chunk so the ores are just below the surface everywhere?

**Comment by tattyseal** [comment]
And then of course ores that use -1 would be at the bottom of the world, the previous example was for 0.

**Comment by tattyseal** [comment]
Or would it be better to do it per-world and keep 0 to the sea level of the world?

**Comment by tattyseal** [comment]
I came up with a new idea. Let the developer decide. When registering an ore to be generated you specify DYNAMIC or SEA_LEVEL. DYNAMIC being the top block in that chunk is 0. SEA_LEVEL being whatever the world sea level is would be 0. Giving the developer control over where they want there ore to generate. And for Structure Generation we have a ChunkGenerateEvent so people can generate structures there.

**Comment by Exuraz** [comment]
I am not a dev or anythin, but how about this for ores and resources;
There is one type of every ore, mods can add a copper ore for example but it will have 1 global ID/type/oreblock handled by Nova. The mod can "enable" this ore  set the texture and all other values, and if there are multiple mods that "enable" the block, the user can choose which values and textures to use. This will prevent the world from being filled with 5 types of copper ore and in theory if user chose, say mod X's texture with mod Y also enabling copper but not changing anything, and removes mod X, the copper will just switch to the texture and settings of mod Y. The same can be done for ingots and dusts, and can add more freedom to different mods in one world without losing items, + the world and inventories will be much cleaner with only 1 type of every ore instead of 5. 

Sorry if it isn't explained very well but it was just a wild idea I got. Right now I have to disable lots of ores to get it to generate only 1 type of copper for instance, which is very annoying.

The same can be done for wood types and other worldgen things, to make cross-game ores and worldgen smoother and cleaner.

**Comment by Cervator** [comment]
I'll re-suggest the link from me posted near the top for the Terasology world gen tutorial, it explains our stuff very well. We have some entirely different concepts to consider, like "facets" that are essentially individual world generation steps that can be put together in a dependency tree so everything happens in the right order, and "borders" so you can request data from an area beyond a single chunk if you need to know about nearby blocks. 

I don't know how you would easily support something like that in a generic API, unless you pick the more advanced system and then can perhaps "dumb it down" for a game that can run purely on "give me a chunk here with these characteristics, all at once" ?

We also have infinite height/depth (cubic chunks), adapted COG thanks to a code donation from JRoush, and even use it to build our caves, which connect perfectly between both horizontal *and* vertical chunks. In the Throughout the Ages world type you can get mountains that are hundreds of meters high (found 1 km+ before) yet still have normal caves and ores and such, while you can also dig down forever (well, to 2 billion blocks down or whatever the data type max is)

Re-linky: https://github.com/Terasology/TutorialWorldGeneration/wiki

**Comment by keybounce** [comment]
Can you add in some sort of overview / "start here" document on that Terasology wiki, to explain how the pieces fit together?

**Comment by Cervator** [comment]
@keybounce Normally I'd be all over that, but world gen is a complex topic to begin with and I think the tutorial wiki covers the pieces very well in the sort of step-by-step approach it takes without overwhelming with terms and definitions :-) The linked pages are meant to be read in order - maybe that could be clearer.

In short "facets" are the features (like: "place a tree built like this in these locations") and can depend on or be depended on by other facets to do things in the right order (like "first determine where to place trees, then dive into placing variants" instead of each tree type being its own thing from start to finish). It allows you to do world gen in a much more structured fashion with fewer conflicts between different systems.

That's the core of the system. Borders just let you extend features into adjacent chunks as needed.

**Comment by Cervator** [comment]
That's a lot of good questions and I might need to defer to one of our world experts to get you the correct technical details. I try to know a little about everything, but rarely everything about any one thing :-)

In short: Yes, I believe our current architecture supports all that, although some parts may still be immature. The tutorial probably doesn't go quite deep enough for your level of interest, the advanced topics are still missing. 

Height tends to be relative to a feature/surface, so you can refer to a height over a particular surface, most typically ground level or sea level, that then later gets placed at the correct absolute height in the world. That way you don't have to worry about getting in the way of some other arbitrarily defined surface like a floating island layer thousands of blocks up in the sky. You could also have DF-style cave layers deep underground with a "surface" of sorts. We have some light regeneration magic to allow a degree of sensible shadow casting without leaving distant enough world below in permanent darkness.

IIRC the border feature can span as many chunks as you need, although naturally the bigger area the more expensive it gets computationally (along with some other fun challenges). The ruins example is perfect for the sort of approach we favor, letting the world features cascade in a tree of sorts. First make some noise, then make sub-layers of noise for specific features, then refine even deeper inside a single specific hit somewhere. Keep going till you're happy.

Currently our worlds tend to originate from a single master module that brings in the types of generators it wants to use, so it knows the order of stuff. Some other self-contained features like cave generation can be optionally plugged in and work fine in at least some cases. I'm unsure if you could eventually run into order issues trying to append enough optional extras.

I'll see if I can poke some world people to get more exact. Might be the easiest to discuss on IRC sometime with the right people on.
